core/templates/dev/head/pages/exploration-editor-page/improvements-tab/services/improvement-modal.service.ts

```diff
@@ -48,6 +48,177 @@ angular.module('oppia').factory('ImprovementModalService', [
+
+              /**
+               * Renders a table for MultipleIncorrectSubmissions issue for a
+               * more friendly UI.
+               */
+              $scope.renderIssueTable = function() {
+                var lars = LearnerActionRenderService;
+                var tableHtml =
+                  lars.renderFinalDisplayBlockForMISIssueHTML(
+                    playthrough.actions.slice(
+                      indexOfActionInFirstDisplayedBlock),
+                    indexOfActionInFirstDisplayedBlock + 1);
+                return tableHtml;
+              };
+
+              /**
+               * Returns the list of learner actions to currently be
+               * displayed in the playthrough modal.
+               * @returns {LearnerAction[]}
+               */
+              $scope.getActionsToRender = function() {
+                if ($scope.issueIsMultipleIncorrectSubmissions) {
+                  // This corresponds to the case where we display the table
+                  // for the issue.
+                  return playthrough.actions.slice(
+                    indexOfFirstDisplayedAction,
+                      indexOfActionInFirstDisplayedBlock);
+                }
+                return playthrough.actions.slice(indexOfFirstDisplayedAction);
+              };
+
+              /**
+               * Returns the index of the learner action.
+               * @param {LearnerAction} learnerAction.
+               * @returns {int}
+               */
+              $scope.getLearnerActionIndex = function(learnerAction) {
+                return playthrough.actions.indexOf(learnerAction) + 1;
+              };
+
+              /**
+               * Computes whether the learner action needs to be highlighted.
+               * @param {LearnerAction} learnerAction.
+               * @returns {boolean}
+               */
+              $scope.isActionHighlighted = function(action) {
+                return $scope.getLearnerActionIndex(action) > (
+                  indexOfActionInFirstDisplayedBlock);
+              };
+
+              /**
+               * Renders the HTML of the learner action.
+               * @param {LearnerAction} learnerAction.
+               * @param {int} actionIndex - The index of the learner action.
+               * @returns {string}
+               */
+              $scope.renderLearnerAction = function(
+                  learnerAction, actionIndex) {
+                return LearnerActionRenderService.renderLearnerAction(
+                  learnerAction, actionIndex);
+              };
+
+
+              /**
+               * Expands the displayed learner actions by the next start
+               * index. If there is only one index, the arrow div is not shown
+               * at all. If the current stop index is the second last one, the
+               * arrow div is hidden after the final stop is reached.
+               */
+              $scope.expandActionsToRender = function() {
```

Comment:
Not sure why you need a separate "shell" function here. Just have a single $scope.expandActions() and that should suffice?

Reply:
Done.